{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "8d30a00c",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-11-12T16:33:20.189782Z",
     "start_time": "2021-11-12T16:33:20.181899Z"
    }
   },
   "source": [
    "# MGE detection"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9f95b25a",
   "metadata": {},
   "source": [
    "Uses the phyletic distribution pattern based mobility score to detect MGEs in genomes."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "96c32472",
   "metadata": {},
   "source": [
    "## Preparation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "3e26e224",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-05-14T08:36:16.642434Z",
     "start_time": "2022-05-14T08:36:14.978627Z"
    }
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from pathlib import Path\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "c6bad3f1",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-05-14T08:36:16.664467Z",
     "start_time": "2022-05-14T08:36:16.657101Z"
    }
   },
   "outputs": [],
   "source": [
    "score_threshold = 0.3\n",
    "n_threshold=10\n",
    "fraction_threshold=0.1\n",
    "n_scc_threshold=5"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "07b05f9c",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-05-14T08:37:38.833343Z",
     "start_time": "2022-05-14T08:37:38.827755Z"
    }
   },
   "outputs": [],
   "source": [
    "project_path = Path().resolve().parent\n",
    "input_path = project_path / \"results\" / \"mobility_files\" /  \"thesis\"\n",
    "output_path = project_path / \"results\" / \"MGE_files\"\n",
    "#os.mkdir(output_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "e0b810e6",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-05-14T08:37:39.720248Z",
     "start_time": "2022-05-14T08:37:39.713461Z"
    }
   },
   "outputs": [],
   "source": [
    "genome_contig=\"GCA_000006865.1-AE005176.1\"\n",
    "genome_contig_file=str(genome_contig+\".csv\")\n",
    "genome2=\"GCA_001868825.1-MLNO01000011.1\"\n",
    "genome2_file=str(genome2+\".csv\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "66bebc0b",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-11-12T16:35:31.207420Z",
     "start_time": "2021-11-12T16:35:31.203766Z"
    }
   },
   "source": [
    "## Load mobility frame"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "b5fea060",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-05-14T08:37:40.589828Z",
     "start_time": "2022-05-14T08:37:40.565478Z"
    }
   },
   "outputs": [],
   "source": [
    "frame=pd.read_csv(input_path / genome2_file, index_col=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "1772f24d",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-05-14T08:37:41.671239Z",
     "start_time": "2022-05-14T08:37:41.644969Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>order</th>\n",
       "      <th>orthogroup</th>\n",
       "      <th>gene</th>\n",
       "      <th>accessory_fraction</th>\n",
       "      <th>tree_score</th>\n",
       "      <th>count</th>\n",
       "      <th>accessory</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>F33765_1</td>\n",
       "      <td>MLNO01000011.1_1</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>F01444_3</td>\n",
       "      <td>MLNO01000011.1_2</td>\n",
       "      <td>0.014493</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>F02943_1</td>\n",
       "      <td>MLNO01000011.1_3</td>\n",
       "      <td>0.260870</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>F00978_1</td>\n",
       "      <td>MLNO01000011.1_4</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>F02552_1</td>\n",
       "      <td>MLNO01000011.1_5</td>\n",
       "      <td>0.730769</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>6</td>\n",
       "      <td>F19307_1</td>\n",
       "      <td>MLNO01000011.1_6</td>\n",
       "      <td>0.523810</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>7</td>\n",
       "      <td>F00386_1</td>\n",
       "      <td>MLNO01000011.1_7</td>\n",
       "      <td>0.694444</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>8</td>\n",
       "      <td>F15837_1</td>\n",
       "      <td>MLNO01000011.1_8</td>\n",
       "      <td>0.666667</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>9</td>\n",
       "      <td>F09060_1</td>\n",
       "      <td>MLNO01000011.1_9</td>\n",
       "      <td>0.622222</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>10</td>\n",
       "      <td>F03841_1</td>\n",
       "      <td>MLNO01000011.1_10</td>\n",
       "      <td>0.234375</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>11</td>\n",
       "      <td>F03189_2</td>\n",
       "      <td>MLNO01000011.1_11</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.924069</td>\n",
       "      <td>2</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>12</td>\n",
       "      <td>F01086_1</td>\n",
       "      <td>MLNO01000011.1_12</td>\n",
       "      <td>0.219512</td>\n",
       "      <td>0.828091</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>13</td>\n",
       "      <td>F00163_6</td>\n",
       "      <td>MLNO01000011.1_13</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.934971</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>14</td>\n",
       "      <td>F01470_3</td>\n",
       "      <td>MLNO01000011.1_14</td>\n",
       "      <td>0.025641</td>\n",
       "      <td>0.924689</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>15</td>\n",
       "      <td>F01341_10</td>\n",
       "      <td>MLNO01000011.1_15</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.947894</td>\n",
       "      <td>2</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>16</td>\n",
       "      <td>F01589_3</td>\n",
       "      <td>MLNO01000011.1_16</td>\n",
       "      <td>0.065789</td>\n",
       "      <td>0.935262</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>17</td>\n",
       "      <td>F00865_3</td>\n",
       "      <td>MLNO01000011.1_17</td>\n",
       "      <td>0.025974</td>\n",
       "      <td>0.944863</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>18</td>\n",
       "      <td>F00867_55</td>\n",
       "      <td>MLNO01000011.1_18</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.833326</td>\n",
       "      <td>2</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>19</td>\n",
       "      <td>F03435_4</td>\n",
       "      <td>MLNO01000011.1_19</td>\n",
       "      <td>0.250000</td>\n",
       "      <td>0.935323</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>20</td>\n",
       "      <td>F02466_2</td>\n",
       "      <td>MLNO01000011.1_20</td>\n",
       "      <td>0.800000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>21</td>\n",
       "      <td>F00467_1</td>\n",
       "      <td>MLNO01000011.1_21</td>\n",
       "      <td>0.093750</td>\n",
       "      <td>0.945653</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>22</td>\n",
       "      <td>F01652_4</td>\n",
       "      <td>MLNO01000011.1_22</td>\n",
       "      <td>0.025316</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>23</td>\n",
       "      <td>F04029_1</td>\n",
       "      <td>MLNO01000011.1_23</td>\n",
       "      <td>0.580645</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    order orthogroup               gene  accessory_fraction  tree_score  \\\n",
       "0       1   F33765_1   MLNO01000011.1_1            0.000000         NaN   \n",
       "1       2   F01444_3   MLNO01000011.1_2            0.014493         NaN   \n",
       "2       3   F02943_1   MLNO01000011.1_3            0.260870         NaN   \n",
       "3       4   F00978_1   MLNO01000011.1_4            0.500000         NaN   \n",
       "4       5   F02552_1   MLNO01000011.1_5            0.730769         NaN   \n",
       "5       6   F19307_1   MLNO01000011.1_6            0.523810         NaN   \n",
       "6       7   F00386_1   MLNO01000011.1_7            0.694444         NaN   \n",
       "7       8   F15837_1   MLNO01000011.1_8            0.666667         NaN   \n",
       "8       9   F09060_1   MLNO01000011.1_9            0.622222         NaN   \n",
       "9      10   F03841_1  MLNO01000011.1_10            0.234375         NaN   \n",
       "10     11   F03189_2  MLNO01000011.1_11            0.000000    0.924069   \n",
       "11     12   F01086_1  MLNO01000011.1_12            0.219512    0.828091   \n",
       "12     13   F00163_6  MLNO01000011.1_13            0.000000    0.934971   \n",
       "13     14   F01470_3  MLNO01000011.1_14            0.025641    0.924689   \n",
       "14     15  F01341_10  MLNO01000011.1_15            0.000000    0.947894   \n",
       "15     16   F01589_3  MLNO01000011.1_16            0.065789    0.935262   \n",
       "16     17   F00865_3  MLNO01000011.1_17            0.025974    0.944863   \n",
       "17     18  F00867_55  MLNO01000011.1_18            0.000000    0.833326   \n",
       "18     19   F03435_4  MLNO01000011.1_19            0.250000    0.935323   \n",
       "19     20   F02466_2  MLNO01000011.1_20            0.800000         NaN   \n",
       "20     21   F00467_1  MLNO01000011.1_21            0.093750    0.945653   \n",
       "21     22   F01652_4  MLNO01000011.1_22            0.025316         NaN   \n",
       "22     23   F04029_1  MLNO01000011.1_23            0.580645         NaN   \n",
       "\n",
       "    count  accessory  \n",
       "0       2      False  \n",
       "1       1      False  \n",
       "2       4      False  \n",
       "3       2      False  \n",
       "4       2      False  \n",
       "5       2      False  \n",
       "6       1      False  \n",
       "7       1      False  \n",
       "8       2      False  \n",
       "9       3      False  \n",
       "10      2      False  \n",
       "11      1      False  \n",
       "12      1      False  \n",
       "13      1      False  \n",
       "14      2      False  \n",
       "15      1      False  \n",
       "16      1      False  \n",
       "17      2      False  \n",
       "18      1      False  \n",
       "19      2      False  \n",
       "20      1      False  \n",
       "21      5      False  \n",
       "22      2      False  "
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "frame"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "417c569e",
   "metadata": {},
   "source": [
    "## Sophisticated MGE detection (front and back boundary detection)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "b086ef85",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-04-25T08:39:27.618834Z",
     "start_time": "2022-04-25T08:39:27.432378Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Possible MGE: 1696-1718\n",
      "Possible MGE: 1541-1555\n",
      "Possible MGE: 1146-1203\n",
      "Possible MGE: 946-964\n",
      "Possible MGE: 858-875\n",
      "Possible MGE: 841-852\n",
      "Possible MGE: 736-804\n",
      "Possible MGE: 500-554\n",
      "Possible MGE: 410-440\n",
      "Possible MGE: 102-119\n"
     ]
    }
   ],
   "source": [
    "###\n",
    "i=0\n",
    "exc=0\n",
    "start_genes=[]\n",
    "end_genes=[]\n",
    "indices=[]\n",
    "lengths=[]\n",
    "for index,row in frame[::-1].iterrows():\n",
    "    consecutive=(row.accessory_fraction>score_threshold)\n",
    "    if consecutive:\n",
    "        i=i+1\n",
    "        #print(i)\n",
    "        #print(index)\n",
    "    else:\n",
    "        exc=exc+1\n",
    "        if(row.accessory|(row['count']>1)|(exc<fraction_threshold*i)):\n",
    "            i=i+1\n",
    "            #print(index)\n",
    "        else:\n",
    "            if i>n_threshold:\n",
    "                indices.append(index)\n",
    "                lengths.append(i)\n",
    "                #print(index)\n",
    "                #start_genes.append(index+1-i)\n",
    "                #end_genes.append(index+1)\n",
    "                print(\"Possible MGE: \" + str(index+2) + \"-\" + str(index+2+i))\n",
    "            i=0\n",
    "            exc=0\n",
    "                \n",
    "# find consecutive genes with mobility score exceeding a threshold "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "d72c745e",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-04-22T09:18:49.050628Z",
     "start_time": "2022-04-22T09:18:49.043790Z"
    }
   },
   "outputs": [],
   "source": [
    "def end_detection(mob_frame):\n",
    "    genes,lengths=get_indices_lengths(mob_frame[::-1])\n",
    "    end_genes=list(np.array(genes)+np.array(lengths))\n",
    "    return [gene+1 for gene in end_genes][::-1], lengths[::-1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "b076c717",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-04-22T09:18:49.305528Z",
     "start_time": "2022-04-22T09:18:49.298561Z"
    }
   },
   "outputs": [],
   "source": [
    "def start_detection(mob_frame):\n",
    "    genes,lengths=get_indices_lengths(mob_frame)\n",
    "    start_genes=list(np.array(genes)-np.array(lengths))\n",
    "    return [gene+1 for gene in start_genes], lengths"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "461dbf03",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-04-22T09:18:49.603925Z",
     "start_time": "2022-04-22T09:18:49.596515Z"
    }
   },
   "outputs": [],
   "source": [
    "def get_indices_lengths(frame, score_threshold=score_threshold, fraction_threshold=fraction_threshold, n_threshold=n_threshold, n_scc_threshold=n_scc_threshold):\n",
    "    i=0\n",
    "    exc=0\n",
    "    n_scc_consecutive=0\n",
    "    indices=[]\n",
    "    lengths=[]\n",
    "    for index,row in frame.iterrows():\n",
    "        #consecutive=(row.accessory_fraction>score_threshold)\n",
    "        if row.accessory_fraction>score_threshold:\n",
    "            # look for consecutive genes with an elevated mobility score\n",
    "            i=i+1\n",
    "        elif row.accessory|(row['count']>1):\n",
    "            #exc are exceptions, so genes that do not have an elevated mobility score\n",
    "            exc=exc+1\n",
    "            # in case these exceptions are multi-copy or accessory, or when they do not appear too close to the start of the potential MGE, they are tolerated\n",
    "            i=i+1\n",
    "        elif ((exc<fraction_threshold*i)&(n_scc_consecutive<n_scc_threshold)):\n",
    "            exc=exc+1\n",
    "            # in case these exceptions do not appear too often and not too often in a consecutive fashion, they are also tolerated\n",
    "            i=i+1\n",
    "            n_scc_consecutive = n_scc_consecutive + 1\n",
    "        elif i>n_threshold:\n",
    "            # if the number of consecutive genes that belong to a potential MGE is large enough, the potential MGE is predicted to be a true MGE\n",
    "            # in this case, the index and length of the MGE are saved\n",
    "            indices.append(index)\n",
    "            lengths.append(i)\n",
    "            # parameters are set to zero again, to find a new (potential) MGE\n",
    "            i=0\n",
    "            exc=0\n",
    "            n_scc_consecutive = 0\n",
    "        else:\n",
    "            # this case captures everything else: genes that are not part of an MGEs\n",
    "            # parameters are set to zero again, to find a new (potential) MGE\n",
    "            i=0\n",
    "            exc=0\n",
    "            n_scc_consecutive = 0\n",
    "    return indices,lengths\n",
    "                "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 131,
   "id": "0d9010fb",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-04-22T11:56:05.094832Z",
     "start_time": "2022-04-22T11:56:05.082978Z"
    }
   },
   "outputs": [],
   "source": [
    "def generate_output(genome_contig_file):\n",
    "    \"\"\"Generates csv files, containing a columns with the contig, a column with a number indicating an individual MGE, and a columns with gene_nrs, indicating the genes that belong to a particular MGE.\"\"\"\n",
    "    contig_extension=genome_contig_file.split('-')[1]\n",
    "    contig=str(contig_extension.split('.')[0]+'.'+contig_extension.split('.')[1])\n",
    "    MGE_list=[]\n",
    "    mob_frame=pd.read_csv(input_path / genome_contig_file, index_col=0)\n",
    "    start_genes, start_lengths=start_detection(mob_frame)\n",
    "    end_genes, end_lengths=end_detection(mob_frame)\n",
    "    MGE=0\n",
    "    # find matching elements in end_genes and start_genes (some potential MGEs will only be detected in one direction)\n",
    "    if (len(start_genes)>0)&(len(end_genes)>0):\n",
    "        for i in range(0,len(start_genes)):\n",
    "            for j in range(0, len(end_genes)):\n",
    "                if (start_genes[i]+start_lengths[i] >= end_genes[j])&(end_genes[j]-end_lengths[j] <= start_genes[i])&(start_genes[i]<end_genes[j]):\n",
    "                    #print(\"length: \"+str(start_lengths[i]))\n",
    "                    #print(\"start: \"+str(start_genes[i]))\n",
    "                    #print(\"end: \"+str(end_genes[j]))\n",
    "                    MGE=MGE+1\n",
    "                    for gene in range(start_genes[i], end_genes[j]+1):\n",
    "                        MGE_list.append({'contig':contig,'MGE':MGE,'gene_nr':gene})\n",
    "                    \n",
    "                \n",
    "    MGE_frame = pd.DataFrame.from_records(MGE_list)\n",
    "    return MGE_frame "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 132,
   "id": "9aa27e02",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-04-22T11:56:06.029173Z",
     "start_time": "2022-04-22T11:56:05.641893Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>contig</th>\n",
       "      <th>MGE</th>\n",
       "      <th>gene_nr</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AE005176.1</td>\n",
       "      <td>1</td>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AE005176.1</td>\n",
       "      <td>1</td>\n",
       "      <td>24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AE005176.1</td>\n",
       "      <td>1</td>\n",
       "      <td>25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AE005176.1</td>\n",
       "      <td>1</td>\n",
       "      <td>26</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AE005176.1</td>\n",
       "      <td>1</td>\n",
       "      <td>27</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>425</th>\n",
       "      <td>AE005176.1</td>\n",
       "      <td>16</td>\n",
       "      <td>2050</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>426</th>\n",
       "      <td>AE005176.1</td>\n",
       "      <td>16</td>\n",
       "      <td>2051</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>427</th>\n",
       "      <td>AE005176.1</td>\n",
       "      <td>16</td>\n",
       "      <td>2052</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>428</th>\n",
       "      <td>AE005176.1</td>\n",
       "      <td>16</td>\n",
       "      <td>2053</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>429</th>\n",
       "      <td>AE005176.1</td>\n",
       "      <td>16</td>\n",
       "      <td>2054</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>430 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         contig  MGE  gene_nr\n",
       "0    AE005176.1    1       23\n",
       "1    AE005176.1    1       24\n",
       "2    AE005176.1    1       25\n",
       "3    AE005176.1    1       26\n",
       "4    AE005176.1    1       27\n",
       "..          ...  ...      ...\n",
       "425  AE005176.1   16     2050\n",
       "426  AE005176.1   16     2051\n",
       "427  AE005176.1   16     2052\n",
       "428  AE005176.1   16     2053\n",
       "429  AE005176.1   16     2054\n",
       "\n",
       "[430 rows x 3 columns]"
      ]
     },
     "execution_count": 132,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "generate_output(genome_contig_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 133,
   "id": "15c4cea5",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-04-22T11:56:06.634377Z",
     "start_time": "2022-04-22T11:56:06.297226Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>contig</th>\n",
       "      <th>MGE</th>\n",
       "      <th>gene_nr</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AE004092.2</td>\n",
       "      <td>1</td>\n",
       "      <td>102</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AE004092.2</td>\n",
       "      <td>1</td>\n",
       "      <td>103</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AE004092.2</td>\n",
       "      <td>1</td>\n",
       "      <td>104</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AE004092.2</td>\n",
       "      <td>1</td>\n",
       "      <td>105</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AE004092.2</td>\n",
       "      <td>1</td>\n",
       "      <td>106</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>AE004092.2</td>\n",
       "      <td>1</td>\n",
       "      <td>107</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>AE004092.2</td>\n",
       "      <td>1</td>\n",
       "      <td>108</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>AE004092.2</td>\n",
       "      <td>1</td>\n",
       "      <td>109</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>AE004092.2</td>\n",
       "      <td>1</td>\n",
       "      <td>110</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>AE004092.2</td>\n",
       "      <td>1</td>\n",
       "      <td>111</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>AE004092.2</td>\n",
       "      <td>1</td>\n",
       "      <td>112</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>AE004092.2</td>\n",
       "      <td>1</td>\n",
       "      <td>113</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>AE004092.2</td>\n",
       "      <td>1</td>\n",
       "      <td>114</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>AE004092.2</td>\n",
       "      <td>1</td>\n",
       "      <td>115</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>AE004092.2</td>\n",
       "      <td>1</td>\n",
       "      <td>116</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>AE004092.2</td>\n",
       "      <td>1</td>\n",
       "      <td>117</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>AE004092.2</td>\n",
       "      <td>1</td>\n",
       "      <td>118</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>AE004092.2</td>\n",
       "      <td>2</td>\n",
       "      <td>411</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>AE004092.2</td>\n",
       "      <td>2</td>\n",
       "      <td>412</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>AE004092.2</td>\n",
       "      <td>2</td>\n",
       "      <td>413</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>AE004092.2</td>\n",
       "      <td>2</td>\n",
       "      <td>414</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>AE004092.2</td>\n",
       "      <td>2</td>\n",
       "      <td>415</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>AE004092.2</td>\n",
       "      <td>2</td>\n",
       "      <td>416</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>AE004092.2</td>\n",
       "      <td>2</td>\n",
       "      <td>417</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>AE004092.2</td>\n",
       "      <td>2</td>\n",
       "      <td>418</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>AE004092.2</td>\n",
       "      <td>2</td>\n",
       "      <td>419</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>AE004092.2</td>\n",
       "      <td>2</td>\n",
       "      <td>420</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>AE004092.2</td>\n",
       "      <td>2</td>\n",
       "      <td>421</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>AE004092.2</td>\n",
       "      <td>2</td>\n",
       "      <td>422</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>AE004092.2</td>\n",
       "      <td>2</td>\n",
       "      <td>423</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>AE004092.2</td>\n",
       "      <td>2</td>\n",
       "      <td>424</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>AE004092.2</td>\n",
       "      <td>2</td>\n",
       "      <td>425</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>AE004092.2</td>\n",
       "      <td>2</td>\n",
       "      <td>426</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>AE004092.2</td>\n",
       "      <td>2</td>\n",
       "      <td>427</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>AE004092.2</td>\n",
       "      <td>2</td>\n",
       "      <td>428</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>AE004092.2</td>\n",
       "      <td>2</td>\n",
       "      <td>429</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>AE004092.2</td>\n",
       "      <td>2</td>\n",
       "      <td>430</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>AE004092.2</td>\n",
       "      <td>2</td>\n",
       "      <td>431</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>AE004092.2</td>\n",
       "      <td>2</td>\n",
       "      <td>432</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>AE004092.2</td>\n",
       "      <td>2</td>\n",
       "      <td>433</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>AE004092.2</td>\n",
       "      <td>2</td>\n",
       "      <td>434</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>AE004092.2</td>\n",
       "      <td>2</td>\n",
       "      <td>435</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>AE004092.2</td>\n",
       "      <td>2</td>\n",
       "      <td>436</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>AE004092.2</td>\n",
       "      <td>2</td>\n",
       "      <td>437</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>AE004092.2</td>\n",
       "      <td>2</td>\n",
       "      <td>438</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>AE004092.2</td>\n",
       "      <td>2</td>\n",
       "      <td>439</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>AE004092.2</td>\n",
       "      <td>3</td>\n",
       "      <td>505</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>AE004092.2</td>\n",
       "      <td>3</td>\n",
       "      <td>506</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>AE004092.2</td>\n",
       "      <td>3</td>\n",
       "      <td>507</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>AE004092.2</td>\n",
       "      <td>3</td>\n",
       "      <td>508</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        contig  MGE  gene_nr\n",
       "0   AE004092.2    1      102\n",
       "1   AE004092.2    1      103\n",
       "2   AE004092.2    1      104\n",
       "3   AE004092.2    1      105\n",
       "4   AE004092.2    1      106\n",
       "5   AE004092.2    1      107\n",
       "6   AE004092.2    1      108\n",
       "7   AE004092.2    1      109\n",
       "8   AE004092.2    1      110\n",
       "9   AE004092.2    1      111\n",
       "10  AE004092.2    1      112\n",
       "11  AE004092.2    1      113\n",
       "12  AE004092.2    1      114\n",
       "13  AE004092.2    1      115\n",
       "14  AE004092.2    1      116\n",
       "15  AE004092.2    1      117\n",
       "16  AE004092.2    1      118\n",
       "17  AE004092.2    2      411\n",
       "18  AE004092.2    2      412\n",
       "19  AE004092.2    2      413\n",
       "20  AE004092.2    2      414\n",
       "21  AE004092.2    2      415\n",
       "22  AE004092.2    2      416\n",
       "23  AE004092.2    2      417\n",
       "24  AE004092.2    2      418\n",
       "25  AE004092.2    2      419\n",
       "26  AE004092.2    2      420\n",
       "27  AE004092.2    2      421\n",
       "28  AE004092.2    2      422\n",
       "29  AE004092.2    2      423\n",
       "30  AE004092.2    2      424\n",
       "31  AE004092.2    2      425\n",
       "32  AE004092.2    2      426\n",
       "33  AE004092.2    2      427\n",
       "34  AE004092.2    2      428\n",
       "35  AE004092.2    2      429\n",
       "36  AE004092.2    2      430\n",
       "37  AE004092.2    2      431\n",
       "38  AE004092.2    2      432\n",
       "39  AE004092.2    2      433\n",
       "40  AE004092.2    2      434\n",
       "41  AE004092.2    2      435\n",
       "42  AE004092.2    2      436\n",
       "43  AE004092.2    2      437\n",
       "44  AE004092.2    2      438\n",
       "45  AE004092.2    2      439\n",
       "46  AE004092.2    3      505\n",
       "47  AE004092.2    3      506\n",
       "48  AE004092.2    3      507\n",
       "49  AE004092.2    3      508"
      ]
     },
     "execution_count": 133,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "generate_output(genome2_file)[0:50]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "97c9c1ab",
   "metadata": {},
   "source": [
    "## Sophisticated MGE detection (MGE joining)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "aff4a3d7",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-04-25T08:55:33.403349Z",
     "start_time": "2022-04-25T08:55:33.398835Z"
    }
   },
   "outputs": [],
   "source": [
    "# parameters\n",
    "score_threshold=0.3\n",
    "initial_n_threshold=5\n",
    "n_threshold=10\n",
    "distance_threshold=5\n",
    "fraction_threshold=0.1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "17bad13b",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-05-14T08:40:30.915715Z",
     "start_time": "2022-05-14T08:40:30.898127Z"
    }
   },
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'initial_n_threshold' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m/var/folders/wm/45ntkynd66gd8wnsdr6vw7h40000gn/T/ipykernel_32725/1573154975.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0;32mdef\u001b[0m \u001b[0mdetect_mges_simple\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mframe\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mn_threshold\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0minitial_n_threshold\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      2\u001b[0m     \u001b[0mi\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      3\u001b[0m     \u001b[0mstart_genes\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      4\u001b[0m     \u001b[0mend_genes\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m     \u001b[0;32mfor\u001b[0m \u001b[0mindex\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mrow\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mframe\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0miterrows\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'initial_n_threshold' is not defined"
     ]
    }
   ],
   "source": [
    "def detect_mges_simple(frame, n_threshold=initial_n_threshold):\n",
    "    i=0\n",
    "    start_genes=[]\n",
    "    end_genes=[]\n",
    "    for index,row in frame.iterrows():\n",
    "        consecutive=row.accessory|(row['count']>1)\n",
    "        if consecutive:\n",
    "            i=i+1\n",
    "        else:\n",
    "            if i>n_threshold:\n",
    "                start_genes.append(index+1-i)\n",
    "                end_genes.append(index+1)\n",
    "                i=0\n",
    "    return start_genes, end_genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "id": "6560fb61",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-04-25T09:16:54.118200Z",
     "start_time": "2022-04-25T09:16:54.107749Z"
    }
   },
   "outputs": [],
   "source": [
    "def join_initial_mges(start_genes, end_genes, distance_threshold=distance_threshold, n_threshold=n_threshold):\n",
    "    \n",
    "    #print(len(start_genes))\n",
    "    #print(len(end_genes))\n",
    "    \n",
    "    indices_to_join=[]\n",
    "    for index in range(len(start_genes)):\n",
    "        if index != 0:\n",
    "            if(start_genes[index]<end_genes[index-1]+distance_threshold):\n",
    "                indices_to_join.append(index)\n",
    "    \n",
    "    for index in range(len(indices_to_join))[::-1]:\n",
    "        start_genes.remove(start_genes[index])\n",
    "        end_genes[index]=end_genes[index-1]\n",
    "        end_genes.remove(end_genes[index-1])\n",
    "    #print(len(start_genes))\n",
    "    #print(len(end_genes))\n",
    "        \n",
    "    for index in range(len(start_genes))[::-1]:\n",
    "        if end_genes[index]-start_genes[index]<n_threshold:\n",
    "            start_genes.remove(start_genes[index])\n",
    "            end_genes.remove(end_genes[index])\n",
    "            #print(\"too small\")\n",
    "    #print(len(start_genes))\n",
    "    #print(len(end_genes))\n",
    "    \n",
    "    return start_genes, end_genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "id": "2fafb4d4",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-04-25T09:19:17.703153Z",
     "start_time": "2022-04-25T09:19:17.693863Z"
    }
   },
   "outputs": [],
   "source": [
    "def generate_output_joining(genome_contig_file):\n",
    "    \"\"\"Generates csv files, containing a columns with the contig, a column with a number indicating an individual MGE, and a columns with gene_nrs, indicating the genes that belong to a particular MGE.\"\"\"\n",
    "    contig_extension=genome_contig_file.split('-')[1]\n",
    "    contig=str(contig_extension.split('.')[0]+'.'+contig_extension.split('.')[1])\n",
    "    MGE_list=[]\n",
    "    mob_frame=pd.read_csv(input_path / genome_contig_file, index_col=0)\n",
    "    start_genes, end_genes=detect_mges_simple(frame)\n",
    "    start_genes, end_genes=join_initial_mges(start_genes, end_genes)\n",
    "    MGE=1\n",
    "    for index in range(len(start_genes)):\n",
    "        for gene in range(start_genes[index], end_genes[index]+1):\n",
    "            MGE_list.append({'contig':contig,'MGE':MGE,'gene_nr':gene})\n",
    "        MGE=MGE+1\n",
    "    MGE_frame = pd.DataFrame.from_records(MGE_list)\n",
    "    return MGE_frame "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "7b5676e2",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-04-25T09:16:45.012948Z",
     "start_time": "2022-04-25T09:16:44.687295Z"
    }
   },
   "outputs": [],
   "source": [
    "start_genes, end_genes=detect_mges_simple(frame)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "id": "f6292b97",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-04-25T09:16:45.198976Z",
     "start_time": "2022-04-25T09:16:45.186321Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "11\n",
      "11\n"
     ]
    }
   ],
   "source": [
    "start_genes, end_genes=join_initial_mges(start_genes, end_genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "id": "7455f383",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-04-25T09:16:45.534646Z",
     "start_time": "2022-04-25T09:16:45.526748Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[173, 425, 502, 740, 827, 858, 1047, 1147, 1543, 1633, 1695]"
      ]
     },
     "execution_count": 77,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "start_genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "id": "ac53eae0",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-04-25T09:16:45.959792Z",
     "start_time": "2022-04-25T09:16:45.950855Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[183, 438, 554, 804, 837, 875, 1058, 1203, 1554, 1643, 1718]"
      ]
     },
     "execution_count": 78,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "end_genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "id": "d5c30ed9",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-04-25T09:19:20.223825Z",
     "start_time": "2022-04-25T09:19:20.047268Z"
    }
   },
   "outputs": [],
   "source": [
    "MGE_frame=generate_output_joining(genome_contig_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "id": "98010f32",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-04-25T09:19:20.508617Z",
     "start_time": "2022-04-25T09:19:20.492471Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>contig</th>\n",
       "      <th>MGE</th>\n",
       "      <th>gene_nr</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AE005176.1</td>\n",
       "      <td>1</td>\n",
       "      <td>173</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AE005176.1</td>\n",
       "      <td>1</td>\n",
       "      <td>174</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AE005176.1</td>\n",
       "      <td>1</td>\n",
       "      <td>175</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AE005176.1</td>\n",
       "      <td>1</td>\n",
       "      <td>176</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AE005176.1</td>\n",
       "      <td>1</td>\n",
       "      <td>177</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>283</th>\n",
       "      <td>AE005176.1</td>\n",
       "      <td>11</td>\n",
       "      <td>1714</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>284</th>\n",
       "      <td>AE005176.1</td>\n",
       "      <td>11</td>\n",
       "      <td>1715</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>285</th>\n",
       "      <td>AE005176.1</td>\n",
       "      <td>11</td>\n",
       "      <td>1716</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>286</th>\n",
       "      <td>AE005176.1</td>\n",
       "      <td>11</td>\n",
       "      <td>1717</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>287</th>\n",
       "      <td>AE005176.1</td>\n",
       "      <td>11</td>\n",
       "      <td>1718</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>288 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         contig  MGE  gene_nr\n",
       "0    AE005176.1    1      173\n",
       "1    AE005176.1    1      174\n",
       "2    AE005176.1    1      175\n",
       "3    AE005176.1    1      176\n",
       "4    AE005176.1    1      177\n",
       "..          ...  ...      ...\n",
       "283  AE005176.1   11     1714\n",
       "284  AE005176.1   11     1715\n",
       "285  AE005176.1   11     1716\n",
       "286  AE005176.1   11     1717\n",
       "287  AE005176.1   11     1718\n",
       "\n",
       "[288 rows x 3 columns]"
      ]
     },
     "execution_count": 89,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "MGE_frame"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "23a19b73",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-04-22T11:45:29.274652Z",
     "start_time": "2022-04-22T11:45:29.270022Z"
    }
   },
   "source": [
    "## Simple MGE detection"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "a78b0928",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-05-14T08:48:39.013729Z",
     "start_time": "2022-05-14T08:48:39.004362Z"
    }
   },
   "outputs": [],
   "source": [
    "def detect_mges_simple(frame, n_threshold=n_threshold):\n",
    "    i=0\n",
    "    start_genes=[]\n",
    "    end_genes=[]\n",
    "    for index,row in frame.iterrows():\n",
    "        consecutive=row.accessory|(row['count']>1)\n",
    "        if consecutive:\n",
    "            i=i+1\n",
    "            pos = index+1\n",
    "            print(str(pos) +\" is multi_copy/accessory\")\n",
    "        else:\n",
    "            if i>=n_threshold:\n",
    "                print(str(pos) + \": element saved\")\n",
    "                start_genes.append(index+1-i)\n",
    "                end_genes.append(index+1)\n",
    "                i=0\n",
    "            else:\n",
    "                i=0\n",
    "    print(start_genes)\n",
    "    print(end_genes)\n",
    "    return start_genes, end_genes\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "04df62a5",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-05-14T08:48:39.487485Z",
     "start_time": "2022-05-14T08:48:39.478031Z"
    }
   },
   "outputs": [],
   "source": [
    "def generate_output_simple(genome_contig_file):\n",
    "    \"\"\"Generates csv files, containing a columns with the contig, a column with a number indicating an individual MGE, and a columns with gene_nrs, indicating the genes that belong to a particular MGE.\"\"\"\n",
    "    contig_extension=genome_contig_file.split('-')[1]\n",
    "    contig=str(contig_extension.split('.')[0]+'.'+contig_extension.split('.')[1])\n",
    "    MGE_list=[]\n",
    "    mob_frame=pd.read_csv(input_path / genome_contig_file, index_col=0)\n",
    "    start_genes, end_genes=detect_mges_simple(mob_frame)\n",
    "    for i in range(0,len(start_genes)):\n",
    "        for gene in range(start_genes[i], end_genes[i]+1):\n",
    "            MGE_list.append({'contig':contig,'MGE':i+1,'gene_nr':gene})\n",
    "    MGE_frame = pd.DataFrame.from_records(MGE_list)\n",
    "    return MGE_frame "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "1a5d5faa",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-05-14T08:48:40.237005Z",
     "start_time": "2022-05-14T08:48:40.211618Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1 is multi_copy/accessory\n",
      "3 is multi_copy/accessory\n",
      "4 is multi_copy/accessory\n",
      "5 is multi_copy/accessory\n",
      "6 is multi_copy/accessory\n",
      "9 is multi_copy/accessory\n",
      "10 is multi_copy/accessory\n",
      "11 is multi_copy/accessory\n",
      "15 is multi_copy/accessory\n",
      "18 is multi_copy/accessory\n",
      "20 is multi_copy/accessory\n",
      "22 is multi_copy/accessory\n",
      "23 is multi_copy/accessory\n",
      "[]\n",
      "[]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: []\n",
       "Index: []"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "generate_output_simple(genome2_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 137,
   "id": "7d51322f",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-04-22T11:57:51.478462Z",
     "start_time": "2022-04-22T11:57:51.150521Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>contig</th>\n",
       "      <th>MGE</th>\n",
       "      <th>gene_nr</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AE004092.2</td>\n",
       "      <td>1</td>\n",
       "      <td>92</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AE004092.2</td>\n",
       "      <td>1</td>\n",
       "      <td>93</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AE004092.2</td>\n",
       "      <td>1</td>\n",
       "      <td>94</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AE004092.2</td>\n",
       "      <td>1</td>\n",
       "      <td>95</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AE004092.2</td>\n",
       "      <td>1</td>\n",
       "      <td>96</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>AE004092.2</td>\n",
       "      <td>1</td>\n",
       "      <td>97</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>AE004092.2</td>\n",
       "      <td>1</td>\n",
       "      <td>98</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>AE004092.2</td>\n",
       "      <td>1</td>\n",
       "      <td>99</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>AE004092.2</td>\n",
       "      <td>1</td>\n",
       "      <td>100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>AE004092.2</td>\n",
       "      <td>1</td>\n",
       "      <td>101</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>AE004092.2</td>\n",
       "      <td>1</td>\n",
       "      <td>102</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>AE004092.2</td>\n",
       "      <td>1</td>\n",
       "      <td>103</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>AE004092.2</td>\n",
       "      <td>1</td>\n",
       "      <td>104</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>AE004092.2</td>\n",
       "      <td>1</td>\n",
       "      <td>105</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>AE004092.2</td>\n",
       "      <td>1</td>\n",
       "      <td>106</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>AE004092.2</td>\n",
       "      <td>1</td>\n",
       "      <td>107</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>AE004092.2</td>\n",
       "      <td>1</td>\n",
       "      <td>108</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>AE004092.2</td>\n",
       "      <td>1</td>\n",
       "      <td>109</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>AE004092.2</td>\n",
       "      <td>1</td>\n",
       "      <td>110</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>AE004092.2</td>\n",
       "      <td>1</td>\n",
       "      <td>111</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>AE004092.2</td>\n",
       "      <td>1</td>\n",
       "      <td>112</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>AE004092.2</td>\n",
       "      <td>2</td>\n",
       "      <td>135</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>AE004092.2</td>\n",
       "      <td>2</td>\n",
       "      <td>136</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>AE004092.2</td>\n",
       "      <td>2</td>\n",
       "      <td>137</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>AE004092.2</td>\n",
       "      <td>2</td>\n",
       "      <td>138</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>AE004092.2</td>\n",
       "      <td>2</td>\n",
       "      <td>139</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>AE004092.2</td>\n",
       "      <td>2</td>\n",
       "      <td>140</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>AE004092.2</td>\n",
       "      <td>2</td>\n",
       "      <td>141</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>AE004092.2</td>\n",
       "      <td>2</td>\n",
       "      <td>142</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>AE004092.2</td>\n",
       "      <td>2</td>\n",
       "      <td>143</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>AE004092.2</td>\n",
       "      <td>2</td>\n",
       "      <td>144</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>AE004092.2</td>\n",
       "      <td>2</td>\n",
       "      <td>145</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>AE004092.2</td>\n",
       "      <td>2</td>\n",
       "      <td>146</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>AE004092.2</td>\n",
       "      <td>2</td>\n",
       "      <td>147</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>AE004092.2</td>\n",
       "      <td>2</td>\n",
       "      <td>148</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>AE004092.2</td>\n",
       "      <td>3</td>\n",
       "      <td>165</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>AE004092.2</td>\n",
       "      <td>3</td>\n",
       "      <td>166</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>AE004092.2</td>\n",
       "      <td>3</td>\n",
       "      <td>167</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>AE004092.2</td>\n",
       "      <td>3</td>\n",
       "      <td>168</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>AE004092.2</td>\n",
       "      <td>3</td>\n",
       "      <td>169</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>AE004092.2</td>\n",
       "      <td>3</td>\n",
       "      <td>170</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>AE004092.2</td>\n",
       "      <td>3</td>\n",
       "      <td>171</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>AE004092.2</td>\n",
       "      <td>3</td>\n",
       "      <td>172</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>AE004092.2</td>\n",
       "      <td>3</td>\n",
       "      <td>173</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>AE004092.2</td>\n",
       "      <td>3</td>\n",
       "      <td>174</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>AE004092.2</td>\n",
       "      <td>3</td>\n",
       "      <td>175</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>AE004092.2</td>\n",
       "      <td>3</td>\n",
       "      <td>176</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>AE004092.2</td>\n",
       "      <td>3</td>\n",
       "      <td>177</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>AE004092.2</td>\n",
       "      <td>4</td>\n",
       "      <td>222</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>AE004092.2</td>\n",
       "      <td>4</td>\n",
       "      <td>223</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        contig  MGE  gene_nr\n",
       "0   AE004092.2    1       92\n",
       "1   AE004092.2    1       93\n",
       "2   AE004092.2    1       94\n",
       "3   AE004092.2    1       95\n",
       "4   AE004092.2    1       96\n",
       "5   AE004092.2    1       97\n",
       "6   AE004092.2    1       98\n",
       "7   AE004092.2    1       99\n",
       "8   AE004092.2    1      100\n",
       "9   AE004092.2    1      101\n",
       "10  AE004092.2    1      102\n",
       "11  AE004092.2    1      103\n",
       "12  AE004092.2    1      104\n",
       "13  AE004092.2    1      105\n",
       "14  AE004092.2    1      106\n",
       "15  AE004092.2    1      107\n",
       "16  AE004092.2    1      108\n",
       "17  AE004092.2    1      109\n",
       "18  AE004092.2    1      110\n",
       "19  AE004092.2    1      111\n",
       "20  AE004092.2    1      112\n",
       "21  AE004092.2    2      135\n",
       "22  AE004092.2    2      136\n",
       "23  AE004092.2    2      137\n",
       "24  AE004092.2    2      138\n",
       "25  AE004092.2    2      139\n",
       "26  AE004092.2    2      140\n",
       "27  AE004092.2    2      141\n",
       "28  AE004092.2    2      142\n",
       "29  AE004092.2    2      143\n",
       "30  AE004092.2    2      144\n",
       "31  AE004092.2    2      145\n",
       "32  AE004092.2    2      146\n",
       "33  AE004092.2    2      147\n",
       "34  AE004092.2    2      148\n",
       "35  AE004092.2    3      165\n",
       "36  AE004092.2    3      166\n",
       "37  AE004092.2    3      167\n",
       "38  AE004092.2    3      168\n",
       "39  AE004092.2    3      169\n",
       "40  AE004092.2    3      170\n",
       "41  AE004092.2    3      171\n",
       "42  AE004092.2    3      172\n",
       "43  AE004092.2    3      173\n",
       "44  AE004092.2    3      174\n",
       "45  AE004092.2    3      175\n",
       "46  AE004092.2    3      176\n",
       "47  AE004092.2    3      177\n",
       "48  AE004092.2    4      222\n",
       "49  AE004092.2    4      223"
      ]
     },
     "execution_count": 137,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "generate_output_simple(genome2_file)[0:50]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6270c250",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.0"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": false
  },
  "varInspector": {
   "cols": {
    "lenName": 16,
    "lenType": 16,
    "lenVar": 40
   },
   "kernels_config": {
    "python": {
     "delete_cmd_postfix": "",
     "delete_cmd_prefix": "del ",
     "library": "var_list.py",
     "varRefreshCmd": "print(var_dic_list())"
    },
    "r": {
     "delete_cmd_postfix": ") ",
     "delete_cmd_prefix": "rm(",
     "library": "var_list.r",
     "varRefreshCmd": "cat(var_dic_list()) "
    }
   },
   "types_to_exclude": [
    "module",
    "function",
    "builtin_function_or_method",
    "instance",
    "_Feature"
   ],
   "window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
