# Thesis: Detection of novel mobile genetic elements in bacterial genomes

This code is meant for the detection of mobile genetic elements (MGEs) in bacterial genomes via two complementary comparative genomics strategies.

## Input data
The input data concerning a set of genomes of choice, consist of the following files:
- genomes_metadata.csv: this is a csv file containing the following columns: genome, gtdb_species, gtdb_genus, gtdb_family, quality, checkm_completeness, checkm_contamination, ncbi_isolation_source, ncbi_strain_identifiers, gtdb_representative,species. Only the first two columns are required.
- pangenome.tsv: this is a tsv file containing the following columns: gene_id, genome, orthogroup. Gene_id is a string that is built up as follows: <contig>.<rank>, where contig represents a unique identifier for a contig, and rank represents the order of the genes in this contig. This file can be generated by the SCARAP toolkit.
- \<genome\>.ffn: this is a nucleotide FASTA file for all genes in a genome. This data is required for all genomes.
- \<genome\>.gff: this is a general feature format for a genome. This data is only required for genomes for which the MGEs need to be detected.

